cmake/kokkos_enable_options.cmake

```
@@ -57,14 +57,12 @@ kokkos_enable_option(
 kokkos_enable_option(TESTS OFF "Whether to build the unit tests")
 kokkos_enable_option(BENCHMARKS OFF "Whether to build the benchmarks")
 kokkos_enable_option(EXAMPLES OFF "Whether to build the examples")
-string(TOUPPER "${CMAKE_BUILD_TYPE}" UPPERCASE_CMAKE_BUILD_TYPE)
-if(UPPERCASE_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
-  kokkos_enable_option(DEBUG ON "Whether to activate extra debug features - may increase compile times")
-  kokkos_enable_option(DEBUG_DUALVIEW_MODIFY_CHECK ON "Debug check on dual views")
-else()
-  kokkos_enable_option(DEBUG OFF "Whether to activate extra debug features - may increase compile times")
-  kokkos_enable_option(DEBUG_DUALVIEW_MODIFY_CHECK OFF "Debug check on dual views")
+kokkos_enable_option(DEBUG ON "Whether to activate extra debug features - may increase compile times")
+kokkos_enable_option(DEBUG_DUALVIEW_MODIFY_CHECK ON "Debug check on dual views")
+if(NOT Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK)
+  message(FATAL_ERROR "Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF isn't supported anymore")
```

I am not sure I agree with https://github.com/kokkos/kokkos/pull/8021#pullrequestreview-2794689046, that it grant us the right to error out at configuration time.  I think it is too aggressive.  I would raise a warning instead.  (The message is fine).
```suggestion
  message(DEPRECATION "Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF isn't supported anymore")
```

---

In that case, we would also need to revert the changes to `containers/src/Kokkos_DualView.hpp`. I'll wait until er merged #8036 before making any changes.

---

Why?  I am only suggesting we should not error out and warn that it will be ignored instead.

---

People currently relying on the check to be off (if there even are any) would likely not set the CMake variable explicitly and thus don't run into the `CMake` warning/error anyway but only fail at runtime.

If they are explicitly setting the option to `OFF`, we turn the runtime error into a configure-time error which seems better. Thus, I would either disable it completly (like we do currently) or only deprecate it while changing the default so they have a chance to keep their code working for now. In the latter case, I would use `DEPRECATION` with `Kokkos_ENABLE_DEPRECATED_CODE_4=ON` enabled and `FATAL_ERROR` otherwise.

---

We never disallowed setting KOKKOS_ENABLE_CUDA_LAMBDA etc. so I think it would be consistent to simply not use it anymore.

---

But I'd be ok with the latter (error out if deprecated code is off).

---

Pleasde do
```suggestion
  if(KOKKOS_ENABLE_DEPRECATED_CODE_4)
    message(DEPRECATION "Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF option is being ignored and is not supported anymore")
  else()
    message(FATAL_ERROR "Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK has been removed and is always enabled")
  endif()
```

---

> We never disallowed setting KOKKOS_ENABLE_CUDA_LAMBDA etc. so I think it would be consistent to simply not use it anymore.

We actually error out for`Kokkos_ENABLE_CUDA_LAMBDA=OFF` , see https://github.com/kokkos/kokkos/blob/be496a754c6518e2cbec1027f84f45f423336638/cmake/kokkos_enable_options.cmake#L258

---

The code you are referring to is in the else-branch of a `if(KOKKOS_ENABLE_DEPRECATED_CODE_4)`